pkg/volume/csi/csi_attacher.go

```diff
@@ -0,0 +1,257 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package csi
+
+import (
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+
+	"k8s.io/api/core/v1"
+	storage "k8s.io/api/storage/v1alpha1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/pkg/volume"
+)
+
+type csiAttacher struct {
+	plugin        *csiPlugin
+	k8s           kubernetes.Interface
+	waitSleepTime time.Duration
+}
+
+// volume.Attacher methods
+var _ volume.Attacher = &csiAttacher{}
+
+func (c *csiAttacher) Attach(spec *volume.Spec, nodeName types.NodeName) (string, error) {
+	if spec == nil {
+		glog.Error(log("attacher.Attach missing volume.Spec"))
+		return "", errors.New("missing spec")
+	}
+
+	csiSource, err := getCSISourceFromSpec(spec)
+	if err != nil {
+		glog.Error(log("attacher.Attach failed to get CSI persistent source: %v", err))
+		return "", errors.New("missing CSI persistent volume")
+	}
+
+	pvName := spec.PersistentVolume.GetName()
+	attachID := fmt.Sprintf("pv-%s", hashAttachmentName(csiSource.VolumeHandle, string(nodeName)))
+
+	attachment := &storage.VolumeAttachment{
+		ObjectMeta: meta.ObjectMeta{
+			Name: attachID,
+		},
+		Spec: storage.VolumeAttachmentSpec{
+			NodeName: string(nodeName),
+			Attacher: csiPluginName,
+			Source: storage.VolumeAttachmentSource{
+				PersistentVolumeName: &pvName,
+			},
+		},
+		Status: storage.VolumeAttachmentStatus{Attached: false},
+	}
+
+	attach, err := c.k8s.StorageV1alpha1().VolumeAttachments().Create(attachment)
+	alreadyExist := false
+	if err != nil {
+		if !apierrs.IsAlreadyExists(err) {
+			glog.Error(log("attacher.Attach failed: %v", err))
+			return "", err
+		}
+		alreadyExist = true
+	}
+
+	if alreadyExist {
+		glog.V(4).Info(log("attachment [%v] for volume [%v] already exists (will not be created)", attach.GetName(), csiSource.VolumeHandle))
+	} else {
+		glog.V(4).Info(log("attachment [%v] for volume [%v] created successfully, will start probing for udpates", attach.GetName(), csiSource.VolumeHandle))
+	}
+
+	// go and probe for status update
+	return c.waitForVolumeAttachment(csiSource.VolumeHandle, attachID)
```

Comment:
@saad-ali I thought attachment is considered successful only when VolumeAttachment.Status.xxx is updated to reflect so.  Let's discuss further.

Reply:
Problem is that once the `VolumeAttachment` object is created, someone needs to clean it up, esp in the case that the actual attachment fails. As coded, if `waitForVolumeAttachment(...)` returns an error, then `Attach(...)` fails, and the attach/detach controller will assume no volume is attached. The `VolumeAttachment` object is never cleaned up.

A few ways to handle this:
1. After a VolumeAttachment object is created, assume, as far as the attach/detach controller is concerned, that the volume is attached. You can still `waitForVolumeAttachment(...), but regardless of if that succeeds or fails, you return no error from `Attach(...)`--this way the attach/detach controller knows that it must call detach (which will clean up the object).
2. On failure of `waitForVolumeAttachment(...)` delete the `VolumeAttachment` created by `Attach(...)`, before returning an error for `Attach(...)`. I don't like this approach because it can cause ping-ponging if a particular attach is taking too long.